fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java

```diff
@@ -6125,6 +6133,43 @@ private ChangedTransactionDetail handleRefundTransaction(final LoanTransaction l
         return changedTransactionDetail;
     }
 
+    public void handleChargebackTransaction(final LoanTransaction chargebackTransaction,
+            final LoanLifecycleStateMachine loanLifecycleStateMachine) {
+
+        chargebackTransaction.updateLoan(this);
+
+        if (!chargebackTransaction.isChargeback()) {
+            final String errorMessage = "A transaction of type chargeback was expected but not received.";
+            throw new InvalidLoanTransactionTypeException("transaction", "is.not.a.chargeback.transaction", errorMessage);
+        }
+
+        final LocalDate loanTransactionDate = chargebackTransaction.getTransactionDate();
+        if (loanTransactionDate.isBefore(getDisbursementDate())) {
+            final String errorMessage = "The transaction date cannot be before the loan disbursement date: "
+                    + getApprovedOnDate().toString();
+            throw new InvalidLoanStateTransitionException("transaction", "cannot.be.before.disbursement.date", errorMessage,
+                    loanTransactionDate, getDisbursementDate());
+        }
+
+        if (loanTransactionDate.isAfter(DateUtils.getBusinessLocalDate())) {
+            final String errorMessage = "The transaction date cannot be in the future.";
+            throw new InvalidLoanStateTransitionException("transaction", "cannot.be.a.future.date", errorMessage, loanTransactionDate);
+        }
+
+        if (chargebackTransaction.isNotZero(loanCurrency())) {
+            addLoanTransaction(chargebackTransaction);
+        }
+
+        final LoanRepaymentScheduleTransactionProcessor loanRepaymentScheduleTransactionProcessor = this.transactionProcessorFactory
+                .determineProcessor(this.transactionProcessingStrategy);
+        loanRepaymentScheduleTransactionProcessor.handleChargeback(chargebackTransaction, getCurrency(),
+                getRepaymentScheduleInstallments());
+
+        updateLoanSummarAndStatus();
+
+        loanLifecycleStateMachine.transition(LoanEvent.LOAN_CHARGEBACK, this);
```

Comment:
This transition is at the wrong place.

The `updateLoanSummaryAndStatus` must be the last because it will change the Loan status based on the recalculated balances.

We need to do the `loanLifecycleStateMachine.transition(LoanEvent.LOAN_CHARGEBACK, this);` prior that to get the proper behaviour:
- Loan status to be moved to `ACTIVE`
- Loan balances got recalculated
- Loan status to be changed to `CLOSED (Obligations met)` in the event the loan is fully repaid or Loan status to be changed to `OVERPAID` in the event the loan is overpaid

Reply:
It was solved in this way:

```
        updateLoanSummaryDerivedFields();
        if (!doPostLoanTransactionChecks(chargebackTransaction.getTransactionDate(), loanLifecycleStateMachine)) {
            loanLifecycleStateMachine.transition(LoanEvent.LOAN_CHARGEBACK, this);
        }
```
Why? Because the Status is evaluated only for Overpaid or Repaid in Full, no other cases to support like Charge back